sycl/test-e2e/CompositeDevice/composite_device.cpp

```diff
@@ -0,0 +1,228 @@
+// RUN: %{build} -o %t.out
+// RUN: env ZE_FLAT_DEVICE_HIERARCHY=COMBINED %{run} %t.out
+// RUN: env ZE_FLAT_DEVICE_HIERARCHY=COMPOSITE %{run} %t.out
+// RUN: env ZE_FLAT_DEVICE_HIERARCHY=FLAT %{run} %t.out
+// REQUIRES: level-zero
+
+#include <sycl/sycl.hpp>
+
+#ifdef SYCL_EXT_ONEAPI_COMPOSITE_DEVICE
+
+using namespace sycl::ext::oneapi::experimental;
+
+bool isL0Backend(sycl::backend backend) {
+  return (backend == sycl::backend::ext_oneapi_level_zero);
+}
+
+bool isCombinedMode() {
+  const char *Mode = std::getenv("ZE_FLAT_DEVICE_HIERARCHY");
+  return (Mode != nullptr) && (std::strcmp(Mode, "COMBINED") == 0);
+}
+
+int main() {
+  sycl::queue q;
+  bool IsCombined = isCombinedMode();
+  auto Platforms = sycl::platform::get_platforms();
+
+  // Check that device::get_devices() and platform::get_devices() do not return
+  // composite devices.
+  {
+    auto Devs = sycl::device::get_devices();
+    for (const auto &D : Devs) {
+      assert(!D.has(sycl::aspect::ext_oneapi_is_composite));
+
+      // If ZE_FLAT_DEVICE_HIERARCHY != COMBINED,
+      // sycl::aspect::ext_oneapi_is_component must be false.
+      assert(!(!IsCombined && D.has(sycl::aspect::ext_oneapi_is_component)));
+    }
+
+    for (const auto &P : Platforms) {
+      bool IsL0 = isL0Backend(P.get_backend());
+      if (!IsL0)
+        continue;
+
+      Devs = P.get_devices();
+      for (const auto &D : Devs) {
+        assert(!D.has(sycl::aspect::ext_oneapi_is_composite));
+
+        // If ZE_FLAT_DEVICE_HIERARCHY != COMBINED,
+        // sycl::aspect::ext_oneapi_is_component must be false.
+        assert(!(!IsCombined && D.has(sycl::aspect::ext_oneapi_is_component)));
```

Comment:
Is this exactly the same as above? Can we outline to a lambda predicate and then use `std::all_of`?